# Assignment is below at the end

- https://scikit-learn.org/stable/modules/tree.html
- https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html
- https://scikit-learn.org/stable/modules/generated/sklearn.tree.plot_tree.html

In [458]:
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
plt.rcParams['figure.figsize'] = (20, 6)
plt.rcParams['font.size'] = 14
import pandas as pd

In [459]:
df = pd.read_csv('/Users/meganrahrig/Desktop/adult.data', index_col=False)
golden = pd.read_csv('/Users/meganrahrig/Desktop/adult.test', index_col=False)

In [460]:
df.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  salary  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [461]:
golden.head()

age   workclass  fnlwgt      education  education-num       marital-status  \
0   25     Private  226802           11th              7        Never-married   
1   38     Private   89814        HS-grad              9   Married-civ-spouse   
2   28   Local-gov  336951     Assoc-acdm             12   Married-civ-spouse   
3   44     Private  160323   Some-college             10   Married-civ-spouse   
4   18           ?  103497   Some-college             10        Never-married   

           occupation relationship    race      sex  capital-gain  \
0   Machine-op-inspct    Own-child   Black     Male             0   
1     Farming-fishing      Husband   White     Male             0   
2     Protective-serv      Husband   White     Male             0   
3   Machine-op-inspct      Husband   Black     Male          7688   
4                   ?    Own-child   White   Female             0   

   capital-loss  hours-per-week  native-country   salary  
0             0              40   United-States   <=50K.  
1             0              50   United-States   <=50K.  
2             0              40   United-States    >50K.  
3             0              40   United-States    >50K.  
4             0              30   United-States   <=50K.

In [462]:
df.head()

age          workclass  fnlwgt   education  education-num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  salary  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [463]:
df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'salary'],
      dtype='object')

In [464]:
from sklearn import preprocessing

In [465]:
# Columns we want to transform
transform_columns = ['sex']

#Columns we can't use because non-numerical
non_num_columns = ['workclass', 'education', 'marital-status', 
                     'occupation', 'relationship', 'race', 'sex', 
                     'native-country']

## First let's try using `pandas.get_dummies()` to transform columns

In [466]:
dummies = pd.get_dummies(df[transform_columns])
dummies

sex_ Female  sex_ Male
0            False       True
1            False       True
2            False       True
3            False       True
4             True      False
...            ...        ...
32556         True      False
32557        False       True
32558         True      False
32559        False       True
32560         True      False

[32561 rows x 2 columns]

In [467]:
dummies.shape

(32561, 2)

## sklearn has a similar process for OneHot Encoding features

In [468]:

from sklearn import preprocessing
enc = preprocessing.OrdinalEncoder()
#onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist", sparse=False)
#onehot.fit(df[transform_columns])

onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist", sparse_output=False)
onehot.fit(df[transform_columns])

OneHotEncoder(handle_unknown='infrequent_if_exist', sparse_output=False)

In [469]:
onehot.categories_

[array([' Female', ' Male'], dtype=object)]

In [470]:
sex = onehot.transform(df[transform_columns])
sex

array([[0., 1.],
       [0., 1.],
       [0., 1.],
       ...,
       [1., 0.],
       [0., 1.],
       [1., 0.]])

In [471]:
sex.shape

(32561, 2)

## In addition to OneHot encoding there is Ordinal Encoding 

In [472]:
enc = preprocessing.OrdinalEncoder()
enc.fit(df[["salary"]])
salary = enc.transform(df[["salary"]])
salary

array([[0.],
       [0.],
       [0.],
       ...,
       [0.],
       [0.],
       [1.]])

In [473]:
enc.categories_[0]

array([' <=50K', ' >50K'], dtype=object)

In [474]:
#x = df.copy()

# transformed = pd.get_dummies(df[transform_columns])
#onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist", sparse=False).fit(df[transform_columns])

transformed = onehot.transform(df[transform_columns])
new_cols = list(onehot.categories_[0].flatten())
#df_trans = pd.DataFrame(transformed, columns=new_cols)

#x = pd.concat(
#    [
#        x.drop(non_num_columns, axis=1), 
#        df_trans
#    ], 
#    axis=1,)

#x["salary"] = enc.transform(df[["salary"]])

enc = preprocessing.OrdinalEncoder()
enc.fit(df[["salary"]])

x = df.copy()
x = pd.concat([x.drop(non_num_columns, axis=1),
               pd.get_dummies(df[transform_columns])], axis=1)
x["salary"] = enc.fit_transform(df[["salary"]])

In [475]:
x.head()

age  fnlwgt  education-num  capital-gain  capital-loss  hours-per-week  \
0   39   77516             13          2174             0              40   
1   50   83311             13             0             0              13   
2   38  215646              9             0             0              40   
3   53  234721              7             0             0              40   
4   28  338409             13             0             0              40   

   salary  sex_ Female  sex_ Male  
0     0.0        False       True  
1     0.0        False       True  
2     0.0        False       True  
3     0.0        False       True  
4     0.0         True      False

In [476]:
xt = golden.copy()

transformed = onehot.transform(xt[transform_columns])
new_cols = list(onehot.categories_[0].flatten())
#df_trans = pd.DataFrame(transformed, columns=new_cols)

#x = pd.concat(
#    [
#        xt.drop(non_num_columns, axis=1), 
#        df_trans
#    ], 
#    axis=1,)

#xt["salary"] = enc.fit_transform(golden[["salary"]])

xt = pd.concat([xt.drop(non_num_columns, axis=1),
               pd.get_dummies(xt[transform_columns])], axis=1)
xt["salary"] = enc.fit_transform(xt[["salary"]])

In [477]:
xt.salary.value_counts()

salary
0.0    12435
1.0     3846
Name: count, dtype: int64

In [478]:
enc.categories_

[array([' <=50K.', ' >50K.'], dtype=object)]

In [479]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

#### Choose the model of your preference: DecisionTree or RandomForest

In [480]:
model = RandomForestClassifier(criterion='entropy')

In [481]:
model = DecisionTreeClassifier(criterion='entropy', max_depth=None)

In [482]:
model.fit(x.drop(['fnlwgt','salary'], axis=1), x.salary)

DecisionTreeClassifier(criterion='entropy')

In [483]:
model.tree_.node_count

8339

In [484]:
list(zip(x.drop(['fnlwgt','salary'], axis=1).columns, model.feature_importances_))

[('age', 0.3218529683948487),
 ('education-num', 0.16079072991049642),
 ('capital-gain', 0.22807132063891608),
 ('capital-loss', 0.07871672446292095),
 ('hours-per-week', 0.1549192845272475),
 ('sex_ Female', 0.0018445963774991916),
 ('sex_ Male', 0.05380437568807101)]

In [485]:
list(zip(x.drop(['fnlwgt','salary'], axis=1).columns, model.feature_importances_))

[('age', 0.3218529683948487),
 ('education-num', 0.16079072991049642),
 ('capital-gain', 0.22807132063891608),
 ('capital-loss', 0.07871672446292095),
 ('hours-per-week', 0.1549192845272475),
 ('sex_ Female', 0.0018445963774991916),
 ('sex_ Male', 0.05380437568807101)]

In [486]:
x.drop(['fnlwgt','salary'], axis=1).head()

age  education-num  capital-gain  capital-loss  hours-per-week  \
0   39             13          2174             0              40   
1   50             13             0             0              13   
2   38              9             0             0              40   
3   53              7             0             0              40   
4   28             13             0             0              40   

   sex_ Female  sex_ Male  
0        False       True  
1        False       True  
2        False       True  
3        False       True  
4         True      False

In [487]:
set(x.columns) - set(xt.columns)

set()

In [488]:
list(x.drop('salary', axis=1).columns)

['age',
 'fnlwgt',
 'education-num',
 'capital-gain',
 'capital-loss',
 'hours-per-week',
 'sex_ Female',
 'sex_ Male']

In [489]:
predictions = model.predict(xt.drop(['fnlwgt','salary'], axis=1))
predictionsx = model.predict(x.drop(['fnlwgt','salary'], axis=1))

In [490]:
#from sklearn.metrics import accuracy_score
#from sklearn.metrics import classification_report
#from sklearn.metrics import confusion_matrix
#from sklearn.metrics import auc
#from sklearn.metrics import roc_curve
from sklearn.metrics import (
    accuracy_score, 
    classification_report, 
    confusion_matrix, auc, roc_curve
)



In [491]:
accuracy_score(xt.salary, predictions)

0.8215097352742461

In [492]:
accuracy_score(xt.salary, predictions)

0.8215097352742461

In [493]:
confusion_matrix(xt.salary, predictions)

array([[11471,   964],
       [ 1942,  1904]])

In [494]:
print(classification_report(xt.salary, predictions))

              precision    recall  f1-score   support

         0.0       0.86      0.92      0.89     12435
         1.0       0.66      0.50      0.57      3846

    accuracy                           0.82     16281
   macro avg       0.76      0.71      0.73     16281
weighted avg       0.81      0.82      0.81     16281



In [495]:
print(classification_report(xt.salary, predictions))

              precision    recall  f1-score   support

         0.0       0.86      0.92      0.89     12435
         1.0       0.66      0.50      0.57      3846

    accuracy                           0.82     16281
   macro avg       0.76      0.71      0.73     16281
weighted avg       0.81      0.82      0.81     16281



In [496]:
accuracy_score(x.salary, predictionsx)

0.8955806025613464

In [497]:
confusion_matrix(x.salary, predictionsx)

array([[24097,   623],
       [ 2777,  5064]])

In [498]:
print(classification_report(x.salary, predictionsx))

              precision    recall  f1-score   support

         0.0       0.90      0.97      0.93     24720
         1.0       0.89      0.65      0.75      7841

    accuracy                           0.90     32561
   macro avg       0.89      0.81      0.84     32561
weighted avg       0.90      0.90      0.89     32561



In [499]:
print(classification_report(x.salary, predictionsx))

              precision    recall  f1-score   support

         0.0       0.90      0.97      0.93     24720
         1.0       0.89      0.65      0.75      7841

    accuracy                           0.90     32561
   macro avg       0.89      0.81      0.84     32561
weighted avg       0.90      0.90      0.89     32561



# For the following use the above `adult` dataset. 

# 1. Show the RandomForest outperforms the DecisionTree for a fixed `max_depth` by training using the train set and calculate `precision`, `recall`, `f1`, `confusion matrix` on golden-test set. Start with only numerical features/columns. (age, education-num, capital-gain, capital-loss, hours-per-week) 

In [500]:
import sklearn
sklearn.__version__
#!pip3 install scikit-learn --upgrade
from sklearn import tree
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score, recall_score, f1_score, confusion_matrix

Decision Tree Classifier

In [501]:
model = DecisionTreeClassifier(criterion='entropy', max_depth=5)

In [502]:
model.fit(x.drop(['fnlwgt','salary'], axis=1), x.salary)


DecisionTreeClassifier(criterion='entropy', max_depth=5)

In [503]:
model.tree_.node_count

49

In [504]:
list(zip(x.drop(['fnlwgt','salary'], axis=1).columns, model.feature_importances_))
list(zip(x.drop(['fnlwgt','salary'], axis=1).columns, model.feature_importances_))

[('age', 0.2788779035969668),
 ('education-num', 0.21323241802445903),
 ('capital-gain', 0.3528567441438105),
 ('capital-loss', 0.04235939343563882),
 ('hours-per-week', 0.008425952300125585),
 ('sex_ Female', 0.0),
 ('sex_ Male', 0.10424758849899919)]

In [505]:
x.drop(['fnlwgt','salary'], axis=1).head()
set(x.columns) - set(xt.columns)
list(x.drop('salary', axis=1).columns)

['age',
 'fnlwgt',
 'education-num',
 'capital-gain',
 'capital-loss',
 'hours-per-week',
 'sex_ Female',
 'sex_ Male']

In [506]:
predictions = model.predict(xt.drop(['fnlwgt','salary'], axis=1))
predictionsx = model.predict(x.drop(['fnlwgt','salary'], axis=1))

In [507]:
accuracy_score(xt.salary, predictions)
confusion_matrix(xt.salary, predictions)
print(classification_report(xt.salary, predictions))

              precision    recall  f1-score   support

         0.0       0.85      0.92      0.89     12435
         1.0       0.66      0.49      0.56      3846

    accuracy                           0.82     16281
   macro avg       0.76      0.71      0.73     16281
weighted avg       0.81      0.82      0.81     16281



In [508]:
accuracy_score(x.salary, predictionsx)
confusion_matrix(x.salary, predictionsx)
print(classification_report(x.salary, predictionsx))


              precision    recall  f1-score   support

         0.0       0.86      0.92      0.89     24720
         1.0       0.68      0.51      0.58      7841

    accuracy                           0.82     32561
   macro avg       0.77      0.71      0.73     32561
weighted avg       0.81      0.82      0.81     32561



Recall, F1, and Precision are all slightly higher in the training model, suggesting that this is a slightly better predictor for true positives than the test model.

Random Forest Classifier:

In [509]:
model2 = RandomForestClassifier(criterion='entropy', max_depth=5)

In [510]:
model2.fit(x.drop(['fnlwgt','salary'], axis=1), x.salary)

RandomForestClassifier(criterion='entropy', max_depth=5)

In [511]:
list(zip(x.drop(['fnlwgt','salary'], axis=1).columns, model2.feature_importances_))
list(zip(x.drop(['fnlwgt','salary'], axis=1).columns, model2.feature_importances_))

[('age', 0.2432232207513162),
 ('education-num', 0.2057489059353251),
 ('capital-gain', 0.30552751442611314),
 ('capital-loss', 0.04570008512304568),
 ('hours-per-week', 0.08299611474924237),
 ('sex_ Female', 0.05639413163457343),
 ('sex_ Male', 0.060410027380384104)]

In [512]:
x.drop(['fnlwgt','salary'], axis=1).head()
set(x.columns) - set(xt.columns)
list(x.drop('salary', axis=1).columns)

['age',
 'fnlwgt',
 'education-num',
 'capital-gain',
 'capital-loss',
 'hours-per-week',
 'sex_ Female',
 'sex_ Male']

In [513]:
predictions2 = model2.predict(xt.drop(['fnlwgt','salary'], axis=1))
predictionsx2 = model2.predict(x.drop(['fnlwgt','salary'], axis=1))

In [514]:
accuracy_score(xt.salary, predictions2)
confusion_matrix(xt.salary, predictions2)
print(classification_report(xt.salary, predictions2))

              precision    recall  f1-score   support

         0.0       0.84      0.97      0.90     12435
         1.0       0.80      0.39      0.53      3846

    accuracy                           0.83     16281
   macro avg       0.82      0.68      0.71     16281
weighted avg       0.83      0.83      0.81     16281



In [515]:
accuracy_score(x.salary, predictionsx2)
confusion_matrix(x.salary, predictionsx2)
print(classification_report(x.salary, predictionsx2))


              precision    recall  f1-score   support

         0.0       0.84      0.97      0.90     24720
         1.0       0.82      0.40      0.54      7841

    accuracy                           0.83     32561
   macro avg       0.83      0.69      0.72     32561
weighted avg       0.83      0.83      0.81     32561



Again, here, Recall, F1, and Precision are all slightly higher in the training model, suggesting that this is a slightly better predictor for true positives than the test model.

- Comparing DecisionTree to RandomForest
> Precision is higher on RandomForest for class "1", suggesting RandomForest is better at predicting true positives
> 
> The F1 Score for Accuracy is slightly higher on the RandomForest model, also suggesting that this is the preferred model.
>
> However, DecisionTree has a slightly higher recall score for class "1", though I would say all metrics considered, RandomForest offers a better predictive model.

# 2. Use a RandomForest or DecisionTree and the `adult` dataset, systematically add new columns, one by one, that are non-numerical but converted using the feature-extraction techniques we learned. Using the golden-test set show [`precision`, `recall`, `f1`, `confusion matrix`] for each additional feature added.

Start over, encode 'sex'

In [516]:
df2 = pd.read_csv('/Users/meganrahrig/Desktop/adult.data', index_col=False)

# Columns we want to transform
transform_columns1 = ['sex']

#Columns we can't use because non-numerical
non_num_columns =  ['marital-status', 'occupation',
'relationship','sex', 'native-country']

dummies = pd.get_dummies(df2[transform_columns1])
#dummies

onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist", sparse_output=False)
onehot.fit(df2[transform_columns1])

#onehot.categories_

sex = onehot.transform(df2[transform_columns1])
sex

#sex.shape
enc = preprocessing.OrdinalEncoder()
enc.fit(df2[["salary"]])
salary = enc.transform(df2[["salary"]])
#salary

transformed = onehot.transform(df2[transform_columns1])
new_cols = list(onehot.categories_[0].flatten())

In [517]:
enc = preprocessing.OrdinalEncoder()
enc.fit(df2[["salary"]])

df3 = df2.copy()
df3 = pd.concat([df3.drop(non_num_columns, axis=1),
               pd.get_dummies(df2[transform_columns1])], axis=1)
df3["salary"] = enc.fit_transform(df2[["salary"]])

In [518]:
df3.head()

age          workclass  fnlwgt   education  education-num    race  \
0   39          State-gov   77516   Bachelors             13   White   
1   50   Self-emp-not-inc   83311   Bachelors             13   White   
2   38            Private  215646     HS-grad              9   White   
3   53            Private  234721        11th              7   Black   
4   28            Private  338409   Bachelors             13   Black   

   capital-gain  capital-loss  hours-per-week  salary  sex_ Female  sex_ Male  
0          2174             0              40     0.0        False       True  
1             0             0              13     0.0        False       True  
2             0             0              40     0.0        False       True  
3             0             0              40     0.0        False       True  
4             0             0              40     0.0         True      False

Next, encode 'education' (called df4)

In [519]:
# Columns we want to transform
transform_columns3 = ['education']

#Columns we can't use because non-numerical
non_num_columns =  ['education']

dummies = pd.get_dummies(df3[transform_columns3])
#dummies

onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist", sparse_output=False)
onehot.fit(df3[transform_columns3])

#onehot.categories_

education = onehot.transform(df3[transform_columns3])
education

#sex.shape
enc = preprocessing.OrdinalEncoder()
enc.fit(df3[["salary"]])
salary = enc.transform(df3[["salary"]])
#salary

transformed = onehot.transform(df3[transform_columns3])
new_cols = list(onehot.categories_[0].flatten())


In [520]:
enc = preprocessing.OrdinalEncoder()
enc.fit(df3[["salary"]])

df4 = df3.copy()
df4 = pd.concat([df4.drop(non_num_columns, axis=1),
               pd.get_dummies(df3[transform_columns3])], axis=1)
df4["salary"] = enc.fit_transform(df4[["salary"]])

In [521]:
df4.head()

age          workclass  fnlwgt  education-num    race  capital-gain  \
0   39          State-gov   77516             13   White          2174   
1   50   Self-emp-not-inc   83311             13   White             0   
2   38            Private  215646              9   White             0   
3   53            Private  234721              7   Black             0   
4   28            Private  338409             13   Black             0   

   capital-loss  hours-per-week  salary  sex_ Female  ...  education_ 9th  \
0             0              40     0.0        False  ...           False   
1             0              13     0.0        False  ...           False   
2             0              40     0.0        False  ...           False   
3             0              40     0.0        False  ...           False   
4             0              40     0.0         True  ...           False   

   education_ Assoc-acdm  education_ Assoc-voc  education_ Bachelors  \
0                  False                 False                  True   
1                  False                 False                  True   
2                  False                 False                 False   
3                  False                 False                 False   
4                  False                 False                  True   

   education_ Doctorate  education_ HS-grad  education_ Masters  \
0                 False               False               False   
1                 False               False               False   
2                 False                True               False   
3                 False               False               False   
4                 False               False               False   

   education_ Preschool  education_ Prof-school  education_ Some-college  
0                 False                   False                    False  
1                 False                   False                    False  
2                 False                   False                    False  
3                 False                   False                    False  
4                 False                   False                    False  

[5 rows x 27 columns]

Next, encode 'workclass' (to be called df5)

In [522]:
# Columns we want to transform
transform_column4 = ['workclass']

#Columns we can't use because non-numerical
non_num_columns4 =  ['workclass']

dummies = pd.get_dummies(df4[transform_columns4])
#dummies

onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist", sparse_output=False)
onehot.fit(df4[transform_columns4])

#onehot.categories_

workclass = onehot.transform(df4[transform_columns4])
workclass

#sex.shape
enc = preprocessing.OrdinalEncoder()
enc.fit(df4[["salary"]])
salary = enc.transform(df4[["salary"]])
#salary

transformed = onehot.transform(df4[transform_columns4])
new_cols = list(onehot.categories_[0].flatten())


In [523]:
enc = preprocessing.OrdinalEncoder()
enc.fit(df4[["salary"]])

df5 = df4.copy()
df5 = pd.concat([df5.drop(non_num_columns4, axis=1),
               pd.get_dummies(df4[transform_columns4])], axis=1)
df5["salary"] = enc.fit_transform(df5[["salary"]])

In [524]:
df5.head()

age  fnlwgt  education-num    race  capital-gain  capital-loss  \
0   39   77516             13   White          2174             0   
1   50   83311             13   White             0             0   
2   38  215646              9   White             0             0   
3   53  234721              7   Black             0             0   
4   28  338409             13   Black             0             0   

   hours-per-week  salary  sex_ Female  sex_ Male  ...  \
0              40     0.0        False       True  ...   
1              13     0.0        False       True  ...   
2              40     0.0        False       True  ...   
3              40     0.0        False       True  ...   
4              40     0.0         True      False  ...   

   education_ Some-college  workclass_ ?  workclass_ Federal-gov  \
0                    False         False                   False   
1                    False         False                   False   
2                    False         False                   False   
3                    False         False                   False   
4                    False         False                   False   

   workclass_ Local-gov  workclass_ Never-worked  workclass_ Private  \
0                 False                    False               False   
1                 False                    False               False   
2                 False                    False                True   
3                 False                    False                True   
4                 False                    False                True   

   workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  workclass_ State-gov  \
0                    False                        False                  True   
1                    False                         True                 False   
2                    False                        False                 False   
3                    False                        False                 False   
4                    False                        False                 False   

   workclass_ Without-pay  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  

[5 rows x 35 columns]

Do the same encoding for the golden file

In [525]:
golden2 = pd.read_csv('/Users/meganrahrig/Desktop/adult.test', index_col=False)
xt2 = golden2.copy()

In [526]:
# Columns we want to transform
transform_columns1 = ['sex']

#Columns we can't use because non-numerical
non_num_columns5 =  ['marital-status', 'occupation',
'relationship','sex', 'native-country']

dummies = pd.get_dummies(xt2[transform_columns1])
#dummies

onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist", sparse_output=False)
onehot.fit(xt2[transform_columns1])

#onehot.categories_

sex = onehot.transform(xt2[transform_columns1])
sex

#sex.shape
enc = preprocessing.OrdinalEncoder()
enc.fit(xt2[["salary"]])
salary = enc.transform(xt2[["salary"]])
#salary

transformed = onehot.transform(xt2[transform_columns1])
new_cols = list(onehot.categories_[0].flatten())

In [527]:
enc = preprocessing.OrdinalEncoder()
enc.fit(xt2[["salary"]])

xt3 = xt2.copy()
xt3 = pd.concat([xt3.drop(non_num_columns5, axis=1),
               pd.get_dummies(xt2[transform_columns1])], axis=1)
xt3["salary"] = enc.fit_transform(xt2[["salary"]])

In [528]:
# Columns we want to transform
transform_columns3 = ['education']

#Columns we can't use because non-numerical
non_num_columns =  ['education']

dummies = pd.get_dummies(xt3[transform_columns3])
#dummies

onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist", sparse_output=False)
onehot.fit(xt3[transform_columns3])

#onehot.categories_

education = onehot.transform(xt3[transform_columns3])
education

#sex.shape
enc = preprocessing.OrdinalEncoder()
enc.fit(xt3[["salary"]])
salary = enc.transform(xt3[["salary"]])
#salary

transformed = onehot.transform(xt3[transform_columns3])
new_cols = list(onehot.categories_[0].flatten())


In [529]:
enc = preprocessing.OrdinalEncoder()
enc.fit(xt3[["salary"]])

xt4 = xt3.copy()
xt4 = pd.concat([xt4.drop(non_num_columns, axis=1),
               pd.get_dummies(xt3[transform_columns3])], axis=1)
xt4["salary"] = enc.fit_transform(xt4[["salary"]])

In [530]:
# Columns we want to transform
transform_column4 = ['workclass']

#Columns we can't use because non-numerical
non_num_columns4 =  ['workclass']

dummies = pd.get_dummies(xt4[transform_columns4])
#dummies

onehot = preprocessing.OneHotEncoder(handle_unknown="infrequent_if_exist", sparse_output=False)
onehot.fit(xt4[transform_columns4])

#onehot.categories_

workclass = onehot.transform(xt4[transform_columns4])
workclass

#sex.shape
enc = preprocessing.OrdinalEncoder()
enc.fit(xt4[["salary"]])
salary = enc.transform(xt4[["salary"]])
#salary

transformed = onehot.transform(xt4[transform_columns4])
new_cols = list(onehot.categories_[0].flatten())


In [531]:
enc = preprocessing.OrdinalEncoder()
enc.fit(xt4[["salary"]])

xt5 = xt4.copy()
xt5 = pd.concat([xt5.drop(non_num_columns4, axis=1),
               pd.get_dummies(xt4[transform_columns4])], axis=1)
xt5["salary"] = enc.fit_transform(xt5[["salary"]])

In [551]:
xt3.head()


age   workclass  fnlwgt      education  education-num    race  \
0   25     Private  226802           11th              7   Black   
1   38     Private   89814        HS-grad              9   White   
2   28   Local-gov  336951     Assoc-acdm             12   White   
3   44     Private  160323   Some-college             10   Black   
4   18           ?  103497   Some-college             10   White   

   capital-gain  capital-loss  hours-per-week  salary  sex_ Female  sex_ Male  
0             0             0              40     0.0        False       True  
1             0             0              50     0.0        False       True  
2             0             0              40     1.0        False       True  
3          7688             0              40     1.0        False       True  
4             0             0              30     0.0         True      False

In [552]:
xt4.head()


age   workclass  fnlwgt  education-num    race  capital-gain  capital-loss  \
0   25     Private  226802              7   Black             0             0   
1   38     Private   89814              9   White             0             0   
2   28   Local-gov  336951             12   White             0             0   
3   44     Private  160323             10   Black          7688             0   
4   18           ?  103497             10   White             0             0   

   hours-per-week  salary  sex_ Female  ...  education_ 9th  \
0              40     0.0        False  ...           False   
1              50     0.0        False  ...           False   
2              40     1.0        False  ...           False   
3              40     1.0        False  ...           False   
4              30     0.0         True  ...           False   

   education_ Assoc-acdm  education_ Assoc-voc  education_ Bachelors  \
0                  False                 False                 False   
1                  False                 False                 False   
2                   True                 False                 False   
3                  False                 False                 False   
4                  False                 False                 False   

   education_ Doctorate  education_ HS-grad  education_ Masters  \
0                 False               False               False   
1                 False                True               False   
2                 False               False               False   
3                 False               False               False   
4                 False               False               False   

   education_ Preschool  education_ Prof-school  education_ Some-college  
0                 False                   False                    False  
1                 False                   False                    False  
2                 False                   False                    False  
3                 False                   False                     True  
4                 False                   False                     True  

[5 rows x 27 columns]

In [553]:
xt5.head()

age  fnlwgt  education-num    race  capital-gain  capital-loss  \
0   25  226802              7   Black             0             0   
1   38   89814              9   White             0             0   
2   28  336951             12   White             0             0   
3   44  160323             10   Black          7688             0   
4   18  103497             10   White             0             0   

   hours-per-week  salary  sex_ Female  sex_ Male  ...  \
0              40     0.0        False       True  ...   
1              50     0.0        False       True  ...   
2              40     1.0        False       True  ...   
3              40     1.0        False       True  ...   
4              30     0.0         True      False  ...   

   education_ Some-college  workclass_ ?  workclass_ Federal-gov  \
0                    False         False                   False   
1                    False         False                   False   
2                    False         False                   False   
3                     True         False                   False   
4                     True          True                   False   

   workclass_ Local-gov  workclass_ Never-worked  workclass_ Private  \
0                 False                    False                True   
1                 False                    False                True   
2                  True                    False               False   
3                 False                    False                True   
4                 False                    False               False   

   workclass_ Self-emp-inc  workclass_ Self-emp-not-inc  workclass_ State-gov  \
0                    False                        False                 False   
1                    False                        False                 False   
2                    False                        False                 False   
3                    False                        False                 False   
4                    False                        False                 False   

   workclass_ Without-pay  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False  

[5 rows x 35 columns]

Decision Tree and Random Forests for the 2 new models (df4 and df5):

In [554]:
model3 = tree.DecisionTreeClassifier(criterion='entropy', max_depth=5)
model3.fit(df3.drop(['education','workclass', 'race', 'fnlwgt','salary'], axis=1), df3.salary)
model3.tree_.node_count

49

In [555]:
list(zip(df3.drop(['education','workclass', 'race', 'fnlwgt','salary'], axis=1).columns, model3.feature_importances_))
list(zip(df3.drop(['education','workclass', 'race', 'fnlwgt','salary'], axis=1).columns, model3.feature_importances_))
df3.drop(['education','workclass', 'race', 'fnlwgt','salary'], axis=1).head()
set(df3.columns) - set(xt3.columns)
list(df3.drop('salary', axis=1).columns)
predictions3 = model3.predict(xt3.drop(['education','workclass', 'race', 'fnlwgt','salary'], axis=1))
predictionsx3 = model3.predict(df3.drop(['education','workclass', 'race', 'fnlwgt','salary'], axis=1))

accuracy_score(xt3.salary, predictions3)
confusion_matrix(xt3.salary, predictions3)
print(classification_report(xt3.salary, predictions3))
accuracy_score(df3.salary, predictionsx3)
confusion_matrix(df3.salary, predictionsx3)
print(classification_report(df3.salary, predictionsx3))


              precision    recall  f1-score   support

         0.0       0.85      0.92      0.89     12435
         1.0       0.66      0.49      0.56      3846

    accuracy                           0.82     16281
   macro avg       0.76      0.71      0.73     16281
weighted avg       0.81      0.82      0.81     16281

              precision    recall  f1-score   support

         0.0       0.86      0.92      0.89     24720
         1.0       0.68      0.51      0.58      7841

    accuracy                           0.82     32561
   macro avg       0.77      0.71      0.73     32561
weighted avg       0.81      0.82      0.81     32561



In [557]:
model4 = tree.DecisionTreeClassifier(criterion='entropy', max_depth=5)
model4.fit(df4.drop(['workclass', 'race', 'fnlwgt','salary'], axis=1), df4.salary)
model4.tree_.node_count

49

In [558]:
list(zip(df4.drop(['workclass', 'race', 'fnlwgt','salary'], axis=1).columns, model4.feature_importances_))
list(zip(df4.drop(['workclass', 'race', 'fnlwgt','salary'], axis=1).columns, model4.feature_importances_))
df4.drop(['workclass', 'race', 'fnlwgt','salary'], axis=1).head()
set(df4.columns) - set(xt4.columns)
list(df4.drop('salary', axis=1).columns)
predictions4 = model4.predict(xt4.drop(['workclass', 'race', 'fnlwgt','salary'], axis=1))
predictionsx4 = model4.predict(df4.drop(['workclass', 'race', 'fnlwgt','salary'], axis=1))

accuracy_score(xt4.salary, predictions4)
confusion_matrix(xt4.salary, predictions4)
print(classification_report(xt4.salary, predictions4))
accuracy_score(df4.salary, predictionsx4)
confusion_matrix(df4.salary, predictionsx4)
print(classification_report(df4.salary, predictionsx4))


              precision    recall  f1-score   support

         0.0       0.85      0.92      0.89     12435
         1.0       0.66      0.49      0.56      3846

    accuracy                           0.82     16281
   macro avg       0.76      0.71      0.73     16281
weighted avg       0.81      0.82      0.81     16281

              precision    recall  f1-score   support

         0.0       0.86      0.92      0.89     24720
         1.0       0.68      0.51      0.58      7841

    accuracy                           0.82     32561
   macro avg       0.77      0.72      0.73     32561
weighted avg       0.81      0.82      0.81     32561



In [559]:
model5 = tree.DecisionTreeClassifier(criterion='entropy', max_depth=5)
model5.fit(df5.drop(['race', 'fnlwgt','salary'], axis=1), df5.salary)
model5.tree_.node_count

49

In [560]:
list(zip(df5.drop(['race', 'fnlwgt','salary'], axis=1).columns, model5.feature_importances_))
list(zip(df5.drop(['race', 'fnlwgt','salary'], axis=1).columns, model5.feature_importances_))
df5.drop(['race', 'fnlwgt','salary'], axis=1).head()
set(df5.columns) - set(xt5.columns)
list(df5.drop('salary', axis=1).columns)
predictions5 = model5.predict(xt5.drop(['race', 'fnlwgt','salary'], axis=1))
predictionsx5 = model5.predict(df5.drop(['race', 'fnlwgt','salary'], axis=1))

accuracy_score(xt5.salary, predictions5)
confusion_matrix(xt5.salary, predictions5)
print(classification_report(xt5.salary, predictions5))
accuracy_score(df5.salary, predictionsx5)
confusion_matrix(df5.salary, predictionsx5)
print(classification_report(df5.salary, predictionsx5))


              precision    recall  f1-score   support

         0.0       0.85      0.92      0.89     12435
         1.0       0.66      0.49      0.56      3846

    accuracy                           0.82     16281
   macro avg       0.76      0.71      0.73     16281
weighted avg       0.81      0.82      0.81     16281

              precision    recall  f1-score   support

         0.0       0.86      0.92      0.89     24720
         1.0       0.68      0.51      0.58      7841

    accuracy                           0.82     32561
   macro avg       0.77      0.72      0.73     32561
weighted avg       0.81      0.82      0.81     32561



Golden Test Results:

- Only 'sex' encoded:

-       precision    recall  f1-score   support

         0.0       0.85      0.92      0.89     12435
         1.0       0.66      0.49      0.56      3846
         accuracy                           0.82     16281
         macro avg       0.76      0.71      0.73     1628
         1weighted avg       0.81      0.82      0.81     16281


>
- 'Sex' and 'education' encoded:
-         precision    recall  f1-score   support

         0.0       0.85      0.92      0.89     12435
         1.0       0.66      0.49      0.56      3846

      accuracy                           0.82     16281

      macro avg       0.76      0.71      0.73     16281

      weighted avg       0.81      0.82      0.81     16281


>
- 'Sex', 'education', and 'workclass' encoded:
-        precision    recall  f1-score   support

         0.0       0.85      0.92      0.89     12435
         1.0       0.66      0.49      0.56      3846

        accuracy                           0.82     16281

      macro avg       0.76      0.71      0.73     16281

      weighted avg       0.81      0.82      0.81     16281

I would have expected error to get smaller as we added more numeric variables. I suspect there is an error in my code that is causing each successive model to still report the same precision, recall, f1-score, and support. However, after examining each dataframe, I have not been able to identify the error.